merkle/src/merkle.rs

```diff
@@ -248,6 +252,158 @@ impl<E: Element> Clone for MmapStore<E> {
+        // the consumer shouldn't need to know how big the final `Store` is
+        // going to be. To *temporarily* accommodate this we *mangle* the
+        // received `size` (in a similar way to `MerkleTree::from_iter`) to
+        // expand it to the entire tree.
+        let pow = next_pow2(size);
+        let size = 2 * pow - 1;
+
+        let byte_len = E::byte_len() * size;
+        let file: File = tempfile().expect("couldn't create temp file");
+        file.set_len(byte_len as u64)
+            .expect(&format!("couldn't set len of {}", byte_len));
+
+        DiskMmapStore {
+            store: unsafe { MmapMut::map_mut(&file).expect("couldn't create map_mut") },
+            len: 0,
+            _e: Default::default(),
+            file,
+        }
+    }
+
+    fn new_from_slice(size: usize, data: &[u8]) -> Self {
+        assert_eq!(data.len() % E::byte_len(), 0);
+
+        let mut res = Self::new(size);
+
+        let end = data.len();
+        res.store[..end].copy_from_slice(data);
+        res.len = data.len() / E::byte_len();
+
+        res
+    }
+
+    fn write_at(&mut self, el: E, i: usize) {
+        let b = E::byte_len();
+        self.store[i * b..(i + 1) * b].copy_from_slice(el.as_ref());
+        self.len += 1;
+    }
+
+    fn read_at(&self, i: usize) -> E {
+        let b = E::byte_len();
+        let start = i * b;
+        let end = (i + 1) * b;
+        let len = self.len * b;
+        assert!(start < len, "start out of range {} >= {}", start, len);
+        assert!(end <= len, "end out of range {} > {}", end, len);
+
+        E::from_slice(&self.store[start..end])
+    }
+
+    fn read_range(&self, r: ops::Range<usize>) -> Vec<E> {
+        let b = E::byte_len();
+        let start = r.start * b;
+        let end = r.end * b;
+        let len = self.len * b;
+        assert!(start < len, "start out of range {} >= {}", start, len);
+        assert!(end <= len, "end out of range {} > {}", end, len);
+
+        self.store[start..end]
+            .chunks(b)
+            .map(E::from_slice)
+            .collect()
+    }
+
+    fn len(&self) -> usize {
+        self.len
+    }
+
+    fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    fn push(&mut self, el: E) {
+        let l = self.len;
+        assert!(
+            (l + 1) * E::byte_len() <= self.store.len(),
+            format!(
+                "not enough space, l: {}, E size {}, store len {}",
+                l,
+                E::byte_len(),
+                self.store.len()
+            )
+        );
+
+        self.write_at(el, l);
+    }
+}
+
+impl<E: Element> DiskMmapStore<E> {
+    pub fn new_with_path(size: usize, path: &Path) -> Self {
+        let byte_len = E::byte_len() * size;
+        let file: File = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .open(&path)
+            .unwrap();
+        file.set_len(byte_len as u64).unwrap();
```

Comment:
Why don't we need to expand the size as in `new`? My expectation (could be misunderstanding) is that `new_with_path` and `new` should be doing the same thing, except that `new_with_path` provides an explicit path.

Is your idea that `new_with_path` will require the caller to have explicitly set the length of the file before passing it?

I can see an argument for having a version used for opening existing trees, whose files have already been written, which should not truncate (or perhaps change at all) its underlying file if the supplied `size` doesn't match the computed size for the file length. Perhaps an extra boolean parameter could be used to specify whether it is better to return an error or mutate the file. In the case of new trees with new files (when creating trees), mutating (by extending from 0) would be desirable. But when reading existing trees, the need to mutate probably represents an error.

Reply:
> Why don't we need to expand the size as in `new`?

This is my error, it should be `new_with_path` adjusting the size (which is directly called by the user) and not `new` called by `MerkleTree` with correct size.

Reply:
(But the fact that this *did* work as is means I'm missing something here and one of my assumptions is off, I'll take a closer look.)

Reply:
Ok, to recap, this was a bug, and the size correction should have been in `new_with_path` (instead of `new`), it was working nonetheless as it was because of the temporary `clone` in https://github.com/filecoin-project/merkle_light/pull/15/files#diff-51a9ffca05030231847ab34ac3206a3aR442 (changing to reference now) that trigger the `new` call (and hence the size fix that wasn't in its correct place).

@porcuquine Still, I'm not sure if this fully answers your question, I'll fix it and ping you to recheck.

Reply:
Moved the size adjustment in https://github.com/filecoin-project/merkle_light/pull/15/commits/2fe35c24f5ebe95dfacd5fe057137d8f049eac51 but still couldn't remove the `clone` call (without modifying the entire `MerkleTree`, which might be a welcomed change anyway).

Reply:
> My expectation (could be misunderstanding) is that `new_with_path` and `new` should be doing the same thing, except that `new_with_path` provides an explicit path.

The issue is that one is called internally with the correct knowledge of the MT and the other is exposed to the consumer (without such knowledge).

> Is your idea that `new_with_path` will require the caller to have explicitly set the length of the file before passing it?

Is not my *idea* or original intention, it's what I was able to find to make things work according to the new API, but I marked it in all-caps to signal this isn't right and we'll need a better solution.

Reply:
I think you added code to set the file length, which answer my first original question. Since the remainder were follow-ons based on the guess that you had meant to leave it out, they don't necessarily need answers.